File: src/Java/Perceptron.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package perceptron;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author David
 */
public class Perceptron {
    
    private List<List<Integer>> entries;
    private List<Double> weights;
    private List<Integer> supervisor;
    private double lr;
    
    public Perceptron(){

        this.entries = new ArrayList<>();
        this.weights = new ArrayList<>();
        this.supervisor = new ArrayList<>();
        this.lr = 0.1;
                
        for( int i =0; i < 2; i++ ){
            this.weights.add(0.0);        
        }

    }
    
    
    public void setEntries(ArrayList entries) {
        this.entries = entries;
    }

    public void setWeights(ArrayList weights) {
        this.weights = weights;
    }

    public void setSupervisor(ArrayList supervisor) {
        this.supervisor = supervisor;
    }

    public void setLr(double lr) {
        this.lr = lr;
    }
    
    public int classify(List<Integer> value){
        
        double  sum = 0 ;
        
        for(int i =0; i < value.size(); i++){
            sum += this.weights.get(i) * value.get(i);
        }
         
        int aux = this.stepFunction(sum);
        return aux;
    }
    
    private int stepFunction(double value){
        return value >= 1.0 ? 1 : 0;
    }
    
    public void train(){
        
        int totalError = 1;
        
        while( totalError != 0 ){
            
            totalError = 0;
            
            for(int i =0; i < this.supervisor.size(); i++ ){
                
                int res = this.classify( this.entries.get(i)  );
                
                int error =  Math.abs( this.supervisor.get(i) - (int) res );
                

                totalError += error;
                
                for(int j = 0; j < this.weights.size(); j++){
                    
                    Double aux = this.weights.get(j);
                    this.weights.set(j,  aux + this.calculateNewWeight(error, i, j)) ;
                    
                    System.out.println("Pesos Atualizados "+this.weights.get(j));
                   
                }
                
                System.out.println("Total de erros "+totalError);
                
            }   
        }
    }
    
    
    private double calculateNewWeight(int error, int i, int j){
        return this.lr * this.entries.get(i).get(j) * error;
    }
}

File: C++/Perceptron.cpp
#include <iostream>
#include <vector>
#include <cmath>


class Perceptron{
	
	private:
		
		std::vector< std::vector<int> > entries;
		std::vector<int> supervisor;
		std::vector<double> weights;
		double lr;
		
		
	public: 
		Perceptron(){
			
			this->lr = 0.1;
			
			for(int i =0; i < 2; i++){
				this->weights.push_back(0.0);
			}
		}
		
		void set_entries(std::vector< std::vector<int> > ent){
			this->entries = ent;
		}
		
		void set_weights(std::vector<double> weig){
			this->weights = weig;
		}
		
		void set_supervisor( std::vector<int> s){
			this->supervisor = s;
		}
		
		void set_lr(double lr){
			this->lr = lr;
		}
				int stepFunction(double value){
			return value >= 1 ? 1 : 0;
		}
		
		double calculateNewValue( double prevValue, int errorRate, int posX, int posY ){
			return prevValue + ( this->lr * this->entries[posX][posY] * errorRate );
		}
		
		int classify( std::vector<int> value ){
			
			double sum=0;
			
			for( int i=0; i<value.size(); i++ ){
				sum += this->weights[i] * value[i];
			}
			return this->stepFunction(sum);
		}
		
		
		void train(){
			
			int totalError = 1;
			
			while( totalError != 0 ){
				
				totalError = 0;
				
				for(int i =0; i < this->supervisor.size(); i++){
					
					int resp = this->classify( this->entries[i] );
					
					int error = abs( this->supervisor[i] - resp );
					
					totalError += error;
					
					for( int j = 0; j < this->weights.size(); j++ ){
						double prev = this->weights[j];
						this->weights[j] = this->calculateNewValue(prev, error, i, j);
						
						std::cout << "Pesos atualizados " << this->weights[j] << std::endl;
					}
					std::cout << "Total de erros " << totalError << std::endl;
				}
			}
		}
};
